### PR TITLE
Port moveit_simple_controller_manager

### PR DESCRIPTION
**Pull request by JafarAbdi**
### Description

Please explain the changes you made, including a reference to the related issue if applicable

### Checklist
- [ ] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extend the tutorials / documentation [reference](http://moveit.ros.org/documentation/contributing/)
- [ ] Document API changes relevant to the user in the MIGRATION.md notes
- [ ] Create tests, which fail without this PR [reference](https://ros-planning.github.io/moveit_tutorials/)
- [ ] Include a screenshot if changing a GUI
- [ ] While waiting for someone to review your request, please help review [another open pull request](https://github.com/ros-planning/moveit/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

File: moveit_core/controller_manager/include/moveit/controller_manager/controller_manager.h [hunk 1]
```diff
@@ -183,7 +183,7 @@ class MoveItControllerManager
   {
   }
 
-  virtual void initialize(std::shared_ptr<rclcpp::Node>& node) = 0;
+  virtual void initialize(const rclcpp::Node::SharedPtr& node) = 0;
```
Comment (JafarAbdi) [inline moveit_core/controller_manager/include/moveit/controller_manager/controller_manager.h]: I'll separate this into a stand-alone PR, once I finish the ros_control package

File: moveit_plugins/moveit_simple_controller_manager/CMakeLists.txt [hunk 2]
```diff
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 3.1.3)
+cmake_minimum_required(VERSION 3.10.2) # For IN_LIST policy support
```
Comment (JafarAbdi) [inline moveit_plugins/moveit_simple_controller_manager/CMakeLists.txt]: I increased the version number for IN_LIST policy support, I should remove the comment

File: moveit_plugins/moveit_simple_controller_manager/src/moveit_simple_controller_manager.cpp [hunk 3]
```diff
@@ -35,108 +35,133 @@
...
 {
 class MoveItSimpleControllerManager : public moveit_controller_manager::MoveItControllerManager
 {
 public:
-  MoveItSimpleControllerManager() : node_handle_("~")
+  MoveItSimpleControllerManager() = default;
+
+  ~MoveItSimpleControllerManager() override = default;
+
+  void initialize(const rclcpp::Node::SharedPtr& node) override
```
Comment (JafarAbdi) [inline moveit_plugins/moveit_simple_controller_manager/src/moveit_simple_controller_manager.cpp]: @JafarAbdi Remember to call "initialize" in the trajectory_execution_manager

File: moveit_plugins/moveit_ros_control_interface/CMakeLists.txt [hunk 4]
```diff
@@ -1,101 +1,66 @@
...
-## Add gtest based cpp test target and link libraries
-# catkin_add_gtest(${PROJECT_NAME}-test test/test_canopen_master.cpp)
-# if(TARGET ${PROJECT_NAME}-test)
-#   target_link_libraries(${PROJECT_NAME}-test ${PROJECT_NAME})
-# endif()
-
-## Add folders to be run by python nosetests
-# catkin_add_nosetests(test)
+pluginlib_export_plugin_description_file(moveit_ros_control_interface moveit_core_plugins.xml)
+pluginlib_export_plugin_description_file(moveit_core moveit_ros_control_interface_plugins.xml)
```
Comment (JafarAbdi) [inline moveit_plugins/moveit_ros_control_interface/CMakeLists.txt]: Port `moveit_ros_control_interface_plugins.xml` to ROS2

**Comment by JafarAbdi** [comment]
This PR assumes `controller_lists.yaml` will have the following format (I added multiple controllers just for illustration)

```yaml
/**:
    ros__parameters:
        controller_names:
          - position_joint_trajectory_controller
          - fake_panda_arm_controller
          - fake_hand_controller
          - franka_gripper
        fake_panda_arm_controller:
            joints:
              - panda_joint1
              - panda_joint2
              - panda_joint3
              - panda_joint4
              - panda_joint5
              - panda_joint6
              - panda_joint7
        fake_hand_controller:
            joints:
              - panda_finger_joint1
              - panda_finger_joint2
        position_joint_trajectory_controller:
           action_ns: follow_joint_trajectory
           type: FollowJointTrajectory
           default: true
           joints:
             - panda_joint1
             - panda_joint2
             - panda_joint3
             - panda_joint4
             - panda_joint5
             - panda_joint6
             - panda_joint7
        franka_gripper:
            action_ns: gripper_action
            type: GripperCommand
            default: true
            joints:
                - panda_finger_joint1
                - panda_finger_joint2

```

File: moveit_plugins/moveit_simple_controller_manager/include/moveit_simple_controller_manager/action_based_controller_handle.h [hunk 5]
```diff
@@ -56,9 +57,13 @@ class ActionBasedControllerHandleBase : public moveit_controller_manager::MoveIt
 
   virtual void addJoint(const std::string& name) = 0;
   virtual void getJoints(std::vector<std::string>& joints) = 0;
-  virtual void configure(XmlRpc::XmlRpcValue& /* config */)
```
Comment (JafarAbdi) [inline moveit_plugins/moveit_simple_controller_manager/include/moveit_simple_controller_manager/action_based_controller_handle.h]: Is this needed anymore .?

File: moveit_plugins/moveit_simple_controller_manager/src/follow_joint_trajectory_controller_handle.cpp [hunk 6]
```diff
@@ -36,50 +36,61 @@
...
+      std::bind(&FollowJointTrajectoryControllerHandle::controllerDoneCallback, this, _1);
+  // Send goal
+  controller_action_client_->async_send_goal(goal, send_goal_options);
+  // TODO(JafarAbdi): current_goal_ .?
   done_ = false;
   last_exec_ = moveit_controller_manager::ExecutionStatus::RUNNING;
   return true;
 }
 
-void FollowJointTrajectoryControllerHandle::configure(XmlRpc::XmlRpcValue& config)
```
Comment (JafarAbdi) [inline moveit_plugins/moveit_simple_controller_manager/src/follow_joint_trajectory_controller_handle.cpp]: Is this function needed .? path_tolerance, goal_tolerance, and goal_time_tolerance aren't set anywhere !

File: moveit_plugins/moveit_simple_controller_manager/src/follow_joint_trajectory_controller_handle.cpp [hunk 7]
```diff
@@ -120,102 +131,93 @@ const char* errorCodeToMessage(int error_code)
...
-    case control_msgs::FollowJointTrajectoryResult::GOAL_TOLERANCE_VIOLATED:
+    case control_msgs::action::FollowJointTrajectory::Result::GOAL_TOLERANCE_VIOLATED:
       return "GOAL_TOLERANCE_VIOLATED";
     default:
       return "unknown error";
   }
 }
 }  // namespace
 
-void FollowJointTrajectoryControllerHandle::configure(XmlRpc::XmlRpcValue& config, const std::string& config_name,
```
Comment (JafarAbdi) [inline moveit_plugins/moveit_simple_controller_manager/src/follow_joint_trajectory_controller_handle.cpp]: Same for this see the other configure function above

File: moveit_plugins/moveit_simple_controller_manager/include/moveit_simple_controller_manager/action_based_controller_handle.h [hunk 8]
```diff
@@ -56,9 +57,9 @@ class ActionBasedControllerHandleBase : public moveit_controller_manager::MoveIt
...
-  {
-  }
+
+protected:
+  const rclcpp::Logger LOGGER = rclcpp::get_logger("action_based_controller_handle");
```
Comment (JafarAbdi) [inline moveit_plugins/moveit_simple_controller_manager/include/moveit_simple_controller_manager/action_based_controller_handle.h]: Not sure if this's how we want to handle having the LOGGER in a header file

File: moveit_plugins/moveit_simple_controller_manager/include/moveit_simple_controller_manager/action_based_controller_handle.h [hunk 9]
```diff
@@ -70,33 +71,37 @@ template <typename T>
...
 
     if (timeout == 0.0)
     {
-      while (ros::ok() && !controller_action_client_->waitForServer(ros::Duration(5.0)))
+      while (rclcpp::ok() && !controller_action_client_->wait_for_action_server(std::chrono::seconds(5)))
       {
-        ROS_WARN_STREAM_NAMED("ActionBasedController", "Waiting for " << getActionName() << " to come up");
-        ros::Duration(1).sleep();
+        RCLCPP_WARN_STREAM(LOGGER, "Waiting for " << getActionName() << " to come up");
+        rclcpp::sleep_for(std::chrono::seconds(1));
```
Comment (JafarAbdi) [inline moveit_plugins/moveit_simple_controller_manager/include/moveit_simple_controller_manager/action_based_controller_handle.h]: Why we have two sleep functions .?

File: moveit_plugins/moveit_simple_controller_manager/include/moveit_simple_controller_manager/action_based_controller_handle.h [hunk 10]
```diff
@@ -70,33 +71,37 @@ template <typename T>
...
     {
-      while (ros::ok() && !controller_action_client_->waitForServer(ros::Duration(timeout / 3)) && ++attempts < 3)
+      while (rclcpp::ok() &&
+             !controller_action_client_->wait_for_action_server(std::chrono::duration<double>(timeout / 3)) &&
+             ++attempts < 3)
       {
-        ROS_WARN_STREAM_NAMED("ActionBasedController", "Waiting for " << getActionName() << " to come up");
-        ros::Duration(1).sleep();
+        RCLCPP_WARN_STREAM(LOGGER, "Waiting for " << getActionName() << " to come up");
+        rclcpp::sleep_for(std::chrono::seconds(1));
```
Comment (JafarAbdi) [inline moveit_plugins/moveit_simple_controller_manager/include/moveit_simple_controller_manager/action_based_controller_handle.h]: Same

File: moveit_plugins/moveit_simple_controller_manager/src/moveit_simple_controller_manager.cpp [hunk 11]
```diff
@@ -35,131 +35,166 @@
...
     }
-
+    std::vector<std::string> controller_names = controller_names_param.as_string_array();
     /* actually create each controller */
-    for (int i = 0; i < controller_list.size(); ++i)  // NOLINT(modernize-loop-convert)
+    for (const std::string& controller_name : controller_names)
     {
-      if (!isStruct(controller_list[i], { "name", "joints", "action_ns", "type" }))
+      std::map<std::string, rclcpp::Parameter> controller_params;
+      node_->get_parameters(controller_name, controller_params);
```
Comment (JafarAbdi) [inline moveit_plugins/moveit_simple_controller_manager/src/moveit_simple_controller_manager.cpp]: Or `node_->get_node_parameters_interface()->get_parameters_by_prefix(controller_name, controller_params);` .?

`node_->get_parameters` is supposed to be called with map of strings to a type (bool, int, string, ..etc) but it works with having `rclcpp::Parameter` not sure if it's hacky

File: moveit_core/controller_manager/include/moveit/controller_manager/controller_manager.h (hunk 1 above)
Comment (henningkayser) [reply moveit_core/controller_manager/include/moveit/controller_manager/controller_manager.h]: I don't think it's necessary to separate this

File: moveit_plugins/moveit_simple_controller_manager/include/moveit_simple_controller_manager/action_based_controller_handle.h (hunk 8 above)
Comment (henningkayser) [reply moveit_plugins/moveit_simple_controller_manager/include/moveit_simple_controller_manager/action_based_controller_handle.h]: You should be able to just declare it in the header and then define it in the cpp file.

File: moveit_plugins/moveit_simple_controller_manager/src/moveit_simple_controller_manager.cpp [hunk 12]
```diff
@@ -35,131 +35,164 @@
...
 
-const std::string LOGNAME("SimpleControllerManager");
+// TOOD(JafarAbdi): Should be in moveit/utils .?
+/*
+ * Check if the map of parameters contains all the input keys
+ */
+bool isSubParameter(const std::map<std::string, rclcpp::Parameter>& parameters, const std::vector<std::string>& keys)
+{
+  std::vector<std::string> missing;
+  for (const std::string& key : keys)
```
Comment (henningkayser) [inline moveit_plugins/moveit_simple_controller_manager/src/moveit_simple_controller_manager.cpp]: Did you test this? I think the parameter keys would include the full name (not only the parameter name suffix). In general I like the idea, but in the future we will have to switch to declaring all parameters before accessing them, so they can't be missing in the first place. For now we can keep it like this, though.

Comment (JafarAbdi) [reply moveit_plugins/moveit_simple_controller_manager/src/moveit_simple_controller_manager.cpp]: Yep it's working, [here](https://gist.github.com/JafarAbdi/6aee7645628e0e59ce5aff3ff3e2b0fa) an example

File: moveit_plugins/moveit_simple_controller_manager/include/moveit_simple_controller_manager/action_based_controller_handle.h (hunk 9 above)
Comment (JafarAbdi) [reply moveit_plugins/moveit_simple_controller_manager/include/moveit_simple_controller_manager/action_based_controller_handle.h]: I'll remove it, please let me know if you want to _retrieve_ it

File: moveit_plugins/moveit_simple_controller_manager/include/moveit_simple_controller_manager/action_based_controller_handle.h [hunk 13]
```diff
@@ -114,24 +119,20 @@ class ActionBasedControllerHandle : public ActionBasedControllerHandleBase
...
+    auto result_future = controller_action_client_->async_get_result(current_goal_);
     if (controller_action_client_ && !done_)
-      return controller_action_client_->waitForResult(timeout);
-#if 1  // TODO: remove when https://github.com/ros/actionlib/issues/155 is fixed
-    // workaround for actionlib issue: waitForResult() might return before our doneCB finished
-    ros::Time deadline = ros::Time::now() + ros::Duration(0.1);  // limit waiting to 0.1s
-    while (!done_ && ros::ok() && ros::Time::now() < deadline)   // Check the done_ flag explicitly,
-      ros::Duration(0.0001).sleep();                             // which is eventually set in finishControllerExecution
-#endif
+      return (rclcpp::spin_until_future_complete(node_, result_future, timeout.to_chrono<std::chrono::seconds>()) ==
```
[the remark this answers is not in the document]
Comment (JafarAbdi) [reply moveit_plugins/moveit_simple_controller_manager/include/moveit_simple_controller_manager/action_based_controller_handle.h]: Doesn't the previous implementation do the same .? blocking until success or timeout .?

File: moveit_plugins/moveit_simple_controller_manager/include/moveit_simple_controller_manager/action_based_controller_handle.h [hunk 14]
```diff
@@ -160,16 +161,18 @@ class ActionBasedControllerHandle : public ActionBasedControllerHandleBase
       return name_ + "/" + namespace_;
   }
 
-  void finishControllerExecution(const actionlib::SimpleClientGoalState& state)
+  void finishControllerExecution(const rclcpp_action::ResultCode& state)
   {
-    ROS_DEBUG_STREAM_NAMED("ActionBasedController", "Controller " << name_ << " is done with state " << state.toString()
-                                                                  << ": " << state.getText());
-    if (state == actionlib::SimpleClientGoalState::SUCCEEDED)
+    // TODO(JafarAbdi): Should I provide an overloaded operator<< for ResultCode .?
```
[the remark this answers is not in the document]
Comment (JafarAbdi) [reply moveit_plugins/moveit_simple_controller_manager/include/moveit_simple_controller_manager/action_based_controller_handle.h]: I'll keep the current implementation and will open a PR to add the streaming operator for ResultCode

File: moveit_plugins/moveit_simple_controller_manager/include/moveit_simple_controller_manager/follow_joint_trajectory_controller_handle.h [hunk 15]
```diff
@@ -47,32 +49,26 @@ namespace moveit_simple_controller_manager
...
 protected:
-  void configure(XmlRpc::XmlRpcValue& config, const std::string& config_name,
-                 std::vector<control_msgs::JointTolerance>& tolerances);
-  static control_msgs::JointTolerance& getTolerance(std::vector<control_msgs::JointTolerance>& tolerances,
-                                                    const std::string& name);
-
-  void controllerDoneCallback(const actionlib::SimpleClientGoalState& state,
-                              const control_msgs::FollowJointTrajectoryResultConstPtr& result);
-
-  void controllerActiveCallback();
```
[the remark this answers is not in the document]
Comment (JafarAbdi) [reply moveit_plugins/moveit_simple_controller_manager/include/moveit_simple_controller_manager/follow_joint_trajectory_controller_handle.h]: The active is very trivial so I replaced it with a lambda function, the feedback one is empty so there's no need for it [see#1](https://github.com/ros-planning/moveit2/pull/158/files#diff-7fd2d600b1d7a1fa0c53ac3dd47cd536L181) [#2](https://github.com/ros-planning/moveit2/pull/158/files#diff-d985568d52bc9433846d768cd18b1b0aR64)

File: moveit_plugins/moveit_simple_controller_manager/include/moveit_simple_controller_manager/gripper_controller_handle.h [hunk 16]
```diff
@@ -47,44 +47,45 @@ namespace moveit_simple_controller_manager
...
     {
-      ROS_ERROR_NAMED("GripperController", "GripperController requires at least one joint trajectory point.");
+      RCLCPP_ERROR(LOGGER, "GripperController requires at least one joint trajectory point.");
       return false;
     }
 
     if (trajectory.joint_trajectory.points.size() > 1)
     {
-      ROS_DEBUG_STREAM_NAMED("GripperController", "Trajectory: " << trajectory.joint_trajectory);
+      // TODO(JafarAbdi): Add helper function for printing misc. msg
```
[the remark this answers is not in the document]
Comment (JafarAbdi) [reply moveit_plugins/moveit_simple_controller_manager/include/moveit_simple_controller_manager/gripper_controller_handle.h]: Commented the if condition and will a todo to enable it once the msg streaming operator is available

File: moveit_plugins/moveit_simple_controller_manager/include/moveit_simple_controller_manager/gripper_controller_handle.h [hunk 17]
```diff
@@ -123,22 +124,38 @@ class GripperControllerHandle : public ActionBasedControllerHandle<control_msgs:
...
+    // Active callback
+    send_goal_options.goal_response_callback = [this](const auto& /* unused-arg */) {
+      RCLCPP_DEBUG_STREAM(LOGGER, name_ << " started execution");
+    };
+    // Result callback
+    send_goal_options.result_callback =
+        std::bind(&GripperControllerHandle::controllerDoneCallback, this, std::placeholders::_1);
+    // Send goal
+    auto current_goal_future = controller_action_client_->async_send_goal(goal, send_goal_options);
+    if (rclcpp::spin_until_future_complete(node_, current_goal_future) != rclcpp::executor::FutureReturnCode::SUCCESS)
```
[the remark this answers is not in the document]
Comment (JafarAbdi) [reply moveit_plugins/moveit_simple_controller_manager/include/moveit_simple_controller_manager/gripper_controller_handle.h]: I used the same method used in the examples please [see](https://github.com/ros2/examples/blob/master/rclcpp/minimal_action_client/not_composable_with_cancel.cpp)

File: moveit_plugins/moveit_simple_controller_manager/include/moveit_simple_controller_manager/action_based_controller_handle.h [hunk 18]
```diff
@@ -51,7 +51,8 @@ namespace moveit_simple_controller_manager
 class ActionBasedControllerHandleBase : public moveit_controller_manager::MoveItControllerHandle
 {
 public:
-  ActionBasedControllerHandleBase(const std::string& name) : moveit_controller_manager::MoveItControllerHandle(name)
+  ActionBasedControllerHandleBase(const std::string& name, const std::string& logger_name)
+    : moveit_controller_manager::MoveItControllerHandle(name), LOGGER(rclcpp::get_logger(logger_name))
```
Comment (JafarAbdi) [inline moveit_plugins/moveit_simple_controller_manager/include/moveit_simple_controller_manager/action_based_controller_handle.h]: The default constructor is implicitly deleted the only public constructor is the copy constructor, the constructor with string is private with get_logger as a friend to the class

**Comment by JafarAbdi** [comment]
@henningkayser I didn't squash the last commit for you to be able to review it alone, is this how you want me to handle the logger in header file .?

File: moveit_plugins/moveit_simple_controller_manager/src/moveit_simple_controller_manager.cpp (hunk 11 above)
Comment (henningkayser) [reply moveit_plugins/moveit_simple_controller_manager/src/moveit_simple_controller_manager.cpp]: I think it's supposed to work like that

**Pull request merged**